pkg/issuer/acme/dns/azuredns/azuredns.go

```diff
@@ -71,6 +73,31 @@ func NewDNSProviderCredentials(environment, clientID, clientSecret, subscription
 	}, nil
 }
 
+func getWorkloadIdentity(env azure.Environment) (*adal.ServicePrincipalToken, error) {
+
+	awiClientId := os.Getenv("AZURE_CLIENT_ID")
+	awiTenantId := os.Getenv("AZURE_TENANT_ID")
+
+	jwtBytes, err := ioutil.ReadFile(os.Getenv("AZURE_FEDERATED_TOKEN_FILE"))
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get Azure Workload Identity token for file: %v", err)
+	}
+
+	jwt := string(jwtBytes)
+
+	oauthConfig, err := adal.NewOAuthConfig(env.ActiveDirectoryEndpoint, awiTenantId)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve OAuth config: %v", err)
+	}
+
+	spt, err := adal.NewServicePrincipalTokenFromFederatedToken(*oauthConfig, awiClientId, jwt, env.ResourceManagerEndpoint)
```

Comment:
@karlschriek first of all, thanks for your contribution!
I noticed that you implemented a similar patch for external-dns here: https://github.com/kubernetes-sigs/external-dns/pull/2885 with a comment `(although note, it does not deal with ensuring that the refreshed token is read from disk)`. As there have been no changes since July, and the PR sadly stays unreviewed till this day, yesterday I decided it'd be fair to try coming up with my own implementation that would tackle that issue while still being based on `adal`.

So, basically, the federated token is valid for 1h, whereas the AAD token that is received in exchange for federated - for 24h ([docs](https://azure.github.io/azure-workload-identity/docs/topics/service-account-labels-and-annotations.html)). As per my understanding, after ~24h the implicitly called `.Refresh()` method for `spt` would start producing errors, because the `spt` struct would have a secret with stale `jwt`, thus failing to get authorized against `ActiveDirectoryEndpoint`.

`adal` doesn't seem to offer any native support for tracking changes in the file with federated token. The only sane method that I was able to find by looking through the library was taking advantage of `customRefreshFunc`, which is supposed to return a fresh access token or an error if any. To avoid having to reinvent the wheel, I wanted to reuse as many existing methods as possible.
The final logic is the following: whenever an access token is about to expire (or just empty right after `spt` is generated), the library would automatically call a wrapper function that would read the federated token from disk and use it to obtain a new access token by invoking the standard process through explicit call of `.Refresh()` method.
You could take a look at my implementation here: https://github.com/kubernetes-sigs/external-dns/pull/3111/

I'm not very familiar with cert-manager code yet and especially whether it has any issuer re-initialization logic in case of token renewal errors, but I guess we'll need to have the same custom wrapper in place here.
I'm still testing the wrapper in external-dns, but as `customRefreshFunc` is successfully called right after external-dns starts, I don't foresee any issues.

WDYT? And have you had a chance to run the patched version of cert-manager for more than 1 day? (just curious what you'd see in logs and how cert-manager behaves)

@wallrj FYI

Reply:
Update on external-dns: I've run it with my patch for over 24h, the access token got refreshed just fine. That convinces me that it's worth porting the wrapper to cert-manager as well.

Reply:
That seems like a perfectly good solution to me! Because you mention external-dns specifically, we currently work around the refresh issue by just restarting the whole Pod every hour ;)